src/lib/array.js

```diff
@@ -63,6 +65,114 @@ exports.ensureArray = function(out, n) {
+    biguint64: typeof BigUint64Array !== 'undefined' ? BigUint64Array : null
+};
+exports.typedArrays = typedArrays;
+
+
+exports.decodeTypedArraySpec = function(v) {
+    // Assume processed by coerceTypedArraySpec
+    v = coerceTypedArraySpec(v);
+    var T = typedArrays[v.dtype];
+    var buffer;
+    if(v.buffer.constructor === ArrayBuffer) {
+        // Already an ArrayBuffer
+        buffer = v.buffer;
+    } else {
+        // Decode, assuming a string
+        buffer = b64.decode(v.buffer);
+    }
+
+    // Check if 1d shape. If so, we're done
+    if(v.ndims === 1) {
+        // Construct single Typed array over entire buffer
+        return new T(buffer);
+    } else {
+        // Reshape into nested plain arrays with innermost
+        // level containing typed arrays
+        // We could eventually adopt an ndarray library
+
+        // Build cumulative product of dimensions
+        var cumulativeShape = v.shape.map(function(a, i) {
+            return a * (v.shape[i - 1] || 1);
+        });
+
+        // Loop of dimensions in reverse order
+        var nestedArray = [];
+        for(var dimInd = v.ndims - 1; dimInd > 0; dimInd--) {
+            var subArrayLength = v.shape[dimInd];
+            var numSubArrays = cumulativeShape[dimInd - 1];
+            var nextArray = [];
+
+            if(dimInd === v.ndims - 1) {
+                // First time through, we build the
+                // inner most typed arrays
+                for(var typedInd = 0; typedInd < numSubArrays; typedInd++) {
+                    var typedOffset = typedInd * subArrayLength;
+                    nextArray.push(
+                        new T(buffer, typedOffset * T.BYTES_PER_ELEMENT, subArrayLength)
+                    );
+                }
+            } else {
+                // Following times through, build
+                // next layer of nested arrays
+                for(var i = 0; i < numSubArrays; i++) {
+                    var offset = i * subArrayLength;
+                    nextArray.push(nextArray.slice(offset, offset + subArrayLength - 1));
+                }
+            }
+
+            // Update nested array with next nesting level
+            nestedArray = nextArray;
+        }
+
+        return nestedArray;
+    }
+};
+
+function isTypedArraySpec(v) {
+    // Assume v has not passed through
+    return isPlainObject(v) && typedArrays[v.dtype] && v.buffer && (
+        Number.isInteger(v.shape) ||
+        (isArrayOrTypedArray(v.shape) &&
+            v.shape.length > 0 &&
+            v.shape.every(function(d) { return Number.isInteger(d); }))
+    );
+}
+exports.isTypedArraySpec = isTypedArraySpec;
+
+function coerceTypedArraySpec(v) {
+    // Assume isTypedArraySpec passed
+    var coerced = {dtype: v.dtype, buffer: v.buffer};
+
+    // Normalize shape to a list
+    if(Number.isInteger(v.shape)) {
+        coerced.shape = [v.shape];
+    } else {
+        coerced.shape = v.shape;
```

Comment:
It might be a good idea to drop this array support for shape attribute; while it cannot be a typedArray itself.
In that case for n dimensional arrays one could pass an string containing comma (or X) separated integers.

Reply:
At the moment, I am using the shape to build a multi-dimensional array when the length is more than 1. Only the inner most layer of the multi-dimensional array is made up of typed arrays, but I pretty sure it'll be a lot better than nothing.